core/lib/types/src/storage/writes/mod.rs

```diff
@@ -188,6 +188,19 @@ fn prepend_header(compressed_state_diffs: Vec<u8>) -> Vec<u8> {
     res.to_vec()
 }
 
+/// Struct for storing tree writes in DB.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct TreeWrite {
```

Comment:
Dumb question: This struct looks close to `StateDiffRecord`; correspondingly, the `tree_writes` field looks close to `compressed_state_diffs`. Could the logic implemented in this PR be implemented by saving `compressed_state_diffs` early if it's possible to extract from the VM execution, and deserializing `compressed_state_diffs` in metadata calculator? If so, this would solve the issue of data duplication, and would eliminate potential problems associated with using `bincode` serialization (can unexpectedly break in the future).

Reply:
So, using `compressed_state_diffs` won't work because the compression is lossy, it doesn't keep hashed keys for repeated writes. Uncompression can only be performed if having additional context (leaf index => hashed key mapping) and querying this context from DB is exactly what we want to prevent.
As for the bincode, I'm not sure, maybe it makes sense to add explicit methods like `to_bytes` and `from_bytes` and use them instead of bincode, this way it will be easier to catch breaking changes in the future, WDYT?

Reply:
Ah, that's a pity. Explicit byte conversions can subjectively be better (provided the requirement of format stability is explained somewhere), but since we cannot piggyback the existing serialization logic, it could be painful to implement.